h5py/_hl/dataset.py

```diff
@@ -396,6 +396,14 @@ def size(self):
             self._cache_props['size'] = size
         return size
 
+    @property
+    def nbytes(self):
+        """Numpy-style attribute giving the dataset size as the number of bytes"""
+        size = self.size
+        if size is None:  # if we are empty, return None
+            return None
```

Comment:
Or 0? It's nice to have a consistent return type where possible.

Reply:
I was also thinking that, but I was concerned that it doesn't mimic `size`.
As in, NumPy's `size` & `nbytes` returns zero for empty datasets:
```
>>> np.empty((0,)).size
0
>>> np.empty((0,)).nbytes
0
```
But `h5py` returns None for empty
```
>>>h5py.Empty(dtype=np.uint16).size

```
Since I think of `nbytes` as "`size`, but scaled" I figured `None` would make sense.  It seems to me either option is logical but `None` is marginally more consistent.  

```suggestion
            return 0
```

Reply:
Numpy doesn't have anything equivalent to `h5py.Empty`, as far as I know: `np.empty((0,))` makes a 1D array of zero length ('empty' there means memory is allocated but not initialised). Both numpy and HDF5 can represent a 0D dataset with 1 entry, but for some reason HDF5 also has a way to store a 0D object with no entries. I don't really know what it's for.

So, no guidance from numpy. h5py used `shape=None` as a convenient way to create an empty dataset (distinguishing it from `shape=()` which makes a scalar). So it made sense for `dset.shape` to give None back. And I guess `dset.size` followed from there.

I'd be inclined not to extend the special case to `nbytes`, but maybe we should get input from other maintainers. If empty dataset support was new, I might also be in favour of making `size` 0. But changing it after the fact is a tougher sell.

Reply:
OK, as none of the other maintainers have commented on this, I'll call it in favour of 0. We've already got other ways to distinguish an empty dataspace from a regular one where one dimension is 0, so I think consistently returning an int here is better.